extensions/procfs/ProcFs.h

```diff
@@ -0,0 +1,64 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <unistd.h>
+
+#include <unordered_map>
+#include <utility>
+#include <string>
+#include <optional>
+#include <filesystem>
+#include <memory>
+
+#include "ProcessStat.h"
+#include "MemInfo.h"
+#include "CpuStat.h"
+#include "NetDev.h"
+#include "DiskStat.h"
+
+#include "core/logging/LoggerConfiguration.h"
+#include "core/logging/Logger.h"
+
+namespace org::apache::nifi::minifi::extensions::procfs {
+
+class ProcFs {
+  static constexpr const char* DEFAULT_ROOT_PATH = "/proc";
+  static constexpr const char* MEMINFO_FILE = "meminfo";
+  static constexpr const char* STAT_FILE = "stat";
+  static constexpr const char* NET_DEV_FILE = "net/dev";
+  static constexpr const char* DISK_STATS_FILE = "diskstats";
+ public:
+  explicit ProcFs(std::filesystem::path path = DEFAULT_ROOT_PATH)
+      : root_path_(std::move(path)) {
+  }
+
+  [[nodiscard]] std::unordered_map<pid_t, ProcessStat> getProcessStats() const;
+  [[nodiscard]] std::unordered_map<std::string, CpuStatData> getCpuStats() const;
+  [[nodiscard]] std::optional<MemInfo> getMemInfo() const;
+  [[nodiscard]] std::unordered_map<std::string, NetDevData> getNetDevs() const;
+  [[nodiscard]] std::unordered_map<std::string, DiskStatData> getDiskStats() const;
```

Comment:
I don't suppose these JSON files will be read by humans much, but in the few cases when that happens, it would be friendlier to those humans to use ordered `map`s instead of `unordered_map`s, so the CPUs, network interfaces, process IDs etc are not in random order.

Lexicographic ordering is not perfect, as we'll have `cpu0, cpu1, cpu11, cpu2, cpu3...`, but it is better than random.

Reply:
If that's a concert, I'd rather go with vector of pairs, to keep the original order. There's also `utils::FlatMap` which does the same AFAIK.

Reply:
I went with a mix of your ideas.
When it comes to CPUStat, DiskStat, NetDev the original order makes sense and the number of entries are expected to be low so I went with FlatMap.
With ProcessStats the original order is already random (due to std::filesystem::directory_iterator) and I can image scenarios where the number of processes go high where the linear lookup of FlatMap could hurt performance in a measurable way, so I went with std::map.

What do you think? https://github.com/apache/nifi-minifi-cpp/pull/1152/commits/d04a5cd15e513d41c8dc9d6ad19a959ddaee0642

Reply:
I like that the orders will be correct, better than with a lexicographic order.

I am a bit bothered by the fact that every time we insert an element into a FlatMap, we do a (linear) lookup first, so inserting N elements takes O(N^2) time.  On the other hand, these Ns are small, so maybe it's not a big issue.  Using a `vector<pair<...>>` instead of a FlatMap would solve this problem, but would make the code less nice.  I would probably go with `vector<pair<...>>` anyway, but it's up to you.

Reply:
I changed the FlatMaps to vector of pairs in https://github.com/apache/nifi-minifi-cpp/pull/1152/commits/da1e32e1bec73705ccad155130947615bb70f804
I also refactored a bit so the processXYZInformation functions are not as overwhelming anymore.

Reply:
looks good, thanks!